### PR TITLE
[BEAM-4752] Add support for newer dill dependency

### PR DESCRIPTION
**Pull request by aaltay**
**Please** add a meaningful description for your change here

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

File: sdks/python/apache_beam/internal/pickler.py [hunk 1]
```diff
@@ -40,6 +40,13 @@
 import dill
 
 
+# Dill 0.28.0 renamed dill.dill to dill._dill:
+# https://github.com/uqfoundation/dill/commit/f0972ecc7a41d0b8acada6042d557068cac69baa
+# TODO: Remove this once Beam depends on dill >= 0.28
+if not hasattr(dill, 'dill'):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/internal/pickler.py]: Can you use getattr is not None instead? (See https://hynek.me/articles/hasattr/)

**Comment by pabloem** [comment]
Retest this please

**Comment by aaltay** [comment]
There are rat errors with unlicensed files, but this PR does not add any new files.

**Comment by aaltay** [comment]
retest this please

**Comment by alanmyrvold** [comment]
I see a test error in https://builds.apache.org/job/beam_PostCommit_Py_ValCont/228

Dataflow logs contain error "No module named _dill"

Could not load main session: Traceback (most recent call last): File "/usr/local/lib/python2.7/site-packages/apache_beam/runners/worker/sdk_worker_main.py", line 125, in main _load_main_session(semi_persistent_directory) File "/usr/local/lib/python2.7/site-packages/apache_beam/runners/worker/sdk_worker_main.py", line 190, in _load_main_session pickler.load_session(session_file) File "/usr/local/lib/python2.7/site-packages/apache_beam/internal/pickler.py", line 260, in load_session return dill.load_session(file_path) File "/usr/local/lib/python2.7/site-packages/dill/dill.py", line 363, in load_session module = unpickler.load() File "/usr/local/lib/python2.7/pickle.py", line 864, in load dispatch[key](self) File "/usr/local/lib/python2.7/pickle.py", line 1096, in load_global klass = self.find_class(module, name) File "/usr/local/lib/python2.7/site-packages/dill/dill.py", line 423, in find_class return StockUnpickler.find_class(self, module, name) File "/usr/local/lib/python2.7/pickle.py", line 1130, in find_class __import__(module) ImportError: No module named _dill Traceback (most recent call last): File "/usr/local/lib/python2.7/site-packages/apache_beam/runners/worker/sdk_worker_main.py", line 125, in main _load_main_session(semi_persistent_directory) File "/usr/local/lib/python2.7/site-packages/apache_beam/runners/worker/sdk_worker_main.py", line 190, in _load_main_session pickler.load_session(session_file) File "/usr/local/lib/python2.7/site-packages/apache_beam/internal/pickler.py", line 260, in load_session return dill.load_session(file_path) File "/usr/local/lib/python2.7/site-packages/dill/dill.py", line 363, in load_session module = unpickler.load() File "/usr/local/lib/python2.7/pickle.py", line 864, in load dispatch[key](self) File "/usr/local/lib/python2.7/pickle.py", line 1096, in load_global klass = self.find_class(module, name) File "/usr/local/lib/python2.7/site-packages/dill/dill.py", line 423, in find_class return StockUnpickler.find_class(self, module, name) File "/usr/local/lib/python2.7/pickle.py", line 1130, in find_class __import__(module) ImportError: No module named _dill

**Comment by aaltay** [comment]
Thank you @alanmyrvold. Sent https://github.com/apache/beam/pull/5946 to rollback.

**Pull request merged**